File: polysolver/runge_kutta.py
"""
Runge-Kutta method for solving
differential equations.


"""

from typing import Callable

def order1(
    fun:Callable,
    x:float,
    y:float,
    z:float,
    h:float
):
    return h*fun(x,y,z)
def order2(
    fun:Callable,
    x:float,
    y:float,
    z:float,
    h:float,
    k1:float,
    l1:float
):
    return h*fun(
        x+0.5*h,
        y+0.5*k1,
        z+0.5*l1
    )
def order3(
    fun:Callable,
    x:float,
    y:float,
    z:float,
    h:float,
    k2:float,
    l2:float,
):
    return h*fun(
        x+0.5*h,
        y+0.5*k2,
        z+0.5*l2
    )
def order4(
    fun:Callable,
    x:float,
    y:float,
    z:float,
    h:float,
    k3:float,
    l3:float,
):
    return h*fun(
        x+h,
        y+k3,
        z+l3
    )

def dy_and_dz(
    yprime:Callable,
    zprime:Callable,
    x:float,
    y:float,
    z:float,
    h:float
):
    k1 = order1(yprime,x,y,z,h)
    l1 = order1(zprime,x,y,z,h)
    k2 = order2(yprime,x,y,z,h,k1,l1)
    l2 = order2(zprime,x,y,z,h,k1,l1)
    k3 = order3(yprime,x,y,z,h,k2,l2)
    l3 = order3(zprime,x,y,z,h,k2,l2)
    k4 = order4(yprime,x,y,z,h,k3,l3)
    l4 = order4(zprime,x,y,z,h,k3,l3)
    return (
        k1/6+k2/3+k3/3+k4/6,
        l1/6+l2/3+l3/3+l4/6
    )

def get_next_xyz(
    yprime:Callable,
    zprime:Callable,
    x:float,
    y:float,
    z:float,
    h:float
):
    dy,dz = dy_and_dz(yprime,zprime,x,y,z,h)
    dx = h
    return x+dx,y+dy,z+dz

File: src/runge_kutta.rs
//// A fourth order Runge-Kutta method

fn order1(fun: &dyn Fn(f64,f64,f64)->f64,x:f64,y:f64,z:f64,h:f64) -> f64 {
    h*fun(x,y,z)
}

fn order2(fun: &dyn Fn(f64,f64,f64)->f64,x:f64,y:f64,z:f64,h:f64,k1:f64,l1:f64) -> f64 {
    h*fun(x+0.5*h,y+0.5*k1,z+0.5*l1)
}
fn order3(fun: &dyn Fn(f64,f64,f64)->f64,x:f64,y:f64,z:f64,h:f64,k2:f64,l2:f64) -> f64 {
    h*fun(x+0.5*h,y+0.5*k2,z+0.5*l2)
}
fn order4(fun: &dyn Fn(f64,f64,f64)->f64,x:f64,y:f64,z:f64,h:f64,k3:f64,l3:f64) -> f64 {
    h*fun(x+h,y+k3,z+l3)
}

fn dy_and_dz(
    yprime: &dyn Fn(f64,f64,f64)->f64,
    zprime: &dyn Fn(f64,f64,f64)->f64,
    x:f64, y:f64, z:f64, h:f64
) -> (f64,f64) {
    let k1 = order1(yprime, x, y, z, h);
    let l1 = order1(zprime, x, y, z, h);
    let k2 = order2(yprime, x, y, z, h, k1, l1);
    let l2 = order2(zprime, x, y, z, h, k1, l1);
    let k3 = order3(yprime, x, y, z, h, k2, l2);
    let l3 = order3(zprime, x, y, z, h, k2, l2);
    let k4 = order4(yprime, x, y, z, h, k3, l3);
    let l4 = order4(zprime, x, y, z, h, k3, l3);
    (
        k1/6.0 + k2/3.0 + k3/3.0 + k4/6.0,
        l1/6.0 + l2/3.0 + l3/3.0 + l4/6.0
    )
}

pub fn get_next_xyz(
    yprime: &dyn Fn(f64,f64,f64)->f64,
    zprime: &dyn Fn(f64,f64,f64)->f64,
    x:f64, y:f64, z:f64, h:f64
) -> (f64,f64,f64) {
    let (dy,dz) = dy_and_dz(yprime, zprime, x, y, z, h);
    (x+h, y+dy, z+dz)
}

#[cfg(test)]
mod tests {
    use super::*;
    fn add(x:f64, y:f64, z:f64) -> f64 {
        x+y+z
    }
    #[test]
    fn test_order1() {
        let k1:f64 = order1(&add, 1.0, 2.0, 3.0, 2.0);
        assert_eq!(k1, 12.0, "k1");
    }

}